### ISSUE TITLE
Regarding Using JavaScript For Logging-In/Registering

### ISSUE THREAD

**Issue by Fawkes-c** [HawkTechnology/WebWareBox#8]
So, we have recently gotten the Login and Registration Pages set-up which is great.

But, the fact that we're going to be using JavaScript/jQuery/AJAX to send the Data over to Flask to handle, is an extremely bad choice. We should have just started doing all of it using Python. It simplifies everything. There are many reasons to not use JavaScript for this System. Let me give the most notable ones though.

1. _Security-Threat_: JS is a threat to online Privacy and Security. It's how Cross-Site-Scripting(**XSS**), is done. It causes Systems in Banks, Governments and Flight-Booking Sites to be easily vulnerable to attacks when JS is used, especially for parts which require Security.

If we use JS for Logging-In and Registration, it may pose a Security-Threat. While I'm not saying it's terrible to use JS for Logging-In(**Tutanota, an encrypted-email service, uses entirely JS**), I do say that we shouldn't use it for the purposes of handling sensitive Data that will be passed around the Site.

2. _JS Is For Front-End_: When I say, "_Front-End_", I'm talking about what the User, the Client sees. Handling sensitive Data such as Emails is meant to be done in the "_Back-End_", which the User, the Client cannot see at all.

As said, JS is meant to be used on the Front-End. It's designed to make Pages more dynamic, functional and add additional features to it. Not for handling sensitive Data. That's why we have PHP, Python, Ruby and Java and *SQL for handling Data.

Again, it is possible to use JS for handling User-Data but I am strongly against it. JS was not designed to be used on the Back-End nor' was it designed to be used to handle sensitive Data. Seeing as we're already using Python for the Back-End, it's essentially best to also use it for handling Data. Python is also a generally more _Secure(**I put "Secure" in Italics as Python does also have vulnerabilities with Security but they're much more rarer than JS vulnerabilities**)_ Language. Also, using Python just simplifies the whole process.

I'm also generally not sure how we're going to use jQuery and AJAX to send Data over to Flask to handle. Any Blog-Posts, Forum Threads or SO(**StackOverflow**) Q&As as to how this will be done will be appreciated.

**Comment by williamdev** [comment]
But then what will we use to send the email and pass to our Python backend?

**Comment by Fawkes-c** [comment]
> But then what will we use to send the email and pass to our Python backend?

That's generally a good question. We'll have to research and look into that. But, it is possible to do this without JS. But, if we can't find of any other way then we will unfortunately have to use JS.

Keep this question open for any further discussion.

**Comment by williamdev** [comment]
I will see once I learn flask

**Comment by williamdev** [comment]
Sorry for the holdup on this one and #4 
I am having issues with registering the email and password to the database. But I will figure it out soon and then we'll have a login and register page without any clientside JS involved, which is cool.

**Comment by Fawkes-c** [comment]
> Sorry for the holdup on this one and #4
> I am having issues with registering the email and password to the database. But I will figure it out soon and then we'll have a login and register page without any clientside JS involved, which is cool.

That's completely fine. I can wait on that. Could you maybe commit the current Code so far and maybe I can inspect it? Cheers.

**Comment by Fawkes-c** [comment]
One question....

**WHAT THE HELL DID YOU DO TO THE FILE-STRUCTURE?**

As soon as I pull in your changes to my Fork, there's a completely new Folder and the `venv/`Directory gets broken for me and I can't activate a Virtual-Environment for Flask to run in. I've literally had to create a new `venv/`, activate it and yet Flask still doesn't run because of things like `from flask_login` and `flask_sqlalchemy`.

Did you follow the official Documentation for Flask? That teaches you how to do everything from Logging-In and connecting to Databases.

I'm sorry, but I cannot test the Software until these things are fixed. Remember, follow the Documentation(**Or as it's sometimes referred to: RTFM - Read The F-ing Manual. And no, just so you know, I'm not trying to be rude or slag you off. I'm just saying for you to read the Documentation for Flask. Besides, I don't think `flask_login` is an actual part of Flask but I may be mistaken**).

**Comment by williamdev** [comment]
All I did was rename it. **You have to install flask_sqlalchemy while inside the virtualenv (after activating with source venv/bin/activate) and just remove the flask_login thing** it's not useful. **I will rename it back to venv if you'd like**. I made the logging in and registering a thing, but I just need login states so that even when the user closes their browser, they can open the WebWareBox back up and then they are still logged in. Maybe I could do this with a database or cookies, but cookies seem unsafe for this since the user can manipulate them to log them into a different account if we don't program it right.

**Comment by williamdev** [comment]
I did not overwrite anything when I renamed to env

**Comment by williamdev** [comment]
I am renaming it back to venv just as a heads up

**Comment by williamdev** [comment]
The functions next to the page routes are used to allow for the submit button on the register and login pages to do their own things. The forms in the login and register pages have actions set that are related to this. For example: <form action="{{ url_for('user_pass_login') }}" method="post"> this is how the forms call the functions inside the backend. Nothing is done in the frontend if we do this. Then the backend does its own thing with the form inputs and checks them against the password hashes inside the database in the login's case. In the register page's case, it makes a record in the AccountDB database with the username and a hashed password. I will rename the env back to venv and then I will remove the imports for modules that aren't used. I just imported those to remind myself that I need those in the long run.

**Comment by williamdev** [comment]
Just install flask_sqlalchemy, it's used for the AccountDB.

**Comment by williamdev** [comment]
If you see errors relating to db.*somethingsomething* in your intellisense/linter or before you run it, ignore those since they do not matter when you run the main.py since I tested main.py out and it worked. Delete the new venv you made and just pull in my changes since the venv folder will have your bin folder as well. Your existing venv folder did not have the Scripts folder which contained activate.bat or activate. I run Windows and the MinGW bash did not work for this. I used the command prompt and then it started working. I ran virtualenv venv, but then I renamed it to env. All I'm saying is that I did not overwrite anything.

**Comment by williamdev** [comment]
Just delete your files that you have cloned from your fork and then pull from dev branch to your fork and then clone your fork to your computer. Things should work since I fixed them.

**Comment by williamdev** [comment]
run virtualenv venv and then run as per the docs after puling in the new chnaged dev branch. if you run virtualenv on windows, like me, it does not create bin folder for linux. if you run virtualenv venv on linux, it does not create scripts folder for windows. so you need to run virtualenv venv to add the bin folder to the prexisting venv so we can run it on both windows and linux

**Comment by Fawkes-c** [comment]
Right. Just figured I'd let you know that I'm currently having issues with Git. First decided to re-fork the Repository and now Git is being a pain in the ass. At this point, I'm ready to lose it all.

I'll try and be quick with this.

**Comment by williamdev** [comment]
Did you make changes that you need to commit? If not, just try cloning the dev branch directly and test that out.
